### ISSUE TITLE
Memory Leak in Microsoft.Extensions.Caching.Memory when handling exceptions

### ISSUE THREAD

**Issue by eerhardt** [dotnet/runtime#42321]
We are leaking objects when calling `MemoryCache.GetOrCreate`  and the factory method is throwing an exception.

### Repro

Run the following application, take a memory snapshot, run for a while and take another snapshot

```C#
using Microsoft.Extensions.Caching.Memory;
using System;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            var cache = new MemoryCache(new MemoryCacheOptions());
            while (true)
            {
                try
                {
                    int value = cache.GetOrCreate("hello", entry =>
                    {
                        throw new Exception();
                        return 10;
                    });
                }
                catch
                {
                    Console.WriteLine("Get Failed");
                }
            }
        }
    }
}
```

![image](https://user-images.githubusercontent.com/8291187/93358615-543d8980-f807-11ea-8c82-93b3bc339a1b.png)

### Analysis

It appears the issue is that we aren't Disposing the CacheEntry instances when an exception is being thrown:

https://github.com/dotnet/runtime/blob/33dba9518b4eb7fbc487fadc9718c408f95a826c/src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs#L98-L112

The reason we aren't calling Dispose is to fix an issue that we were caching `null` when an Exception was being thrown. See https://github.com/aspnet/Caching/pull/216.

Disposing the CacheEntry is important because every time you create a CacheEntry, it gets stuck in an AsyncLocal "stack":

https://github.com/dotnet/runtime/blob/33dba9518b4eb7fbc487fadc9718c408f95a826c/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntryHelper.cs#L28-L36

and disposing it is what "pops" it off the stack:

https://github.com/dotnet/runtime/blob/33dba9518b4eb7fbc487fadc9718c408f95a826c/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntryHelper.cs#L50-L63

We should always be `Disposing` the entry. That way the `ScopeLease` object is always disposed, and the stack is cleared.

However, we still need to fix the original problem: Don't cache `null` when an Exception happens. The reason this happens is because Disposing the CacheEntry is what "commits the entry into the cache". This method gets called from CacheEntry.Dispose:

https://github.com/dotnet/runtime/blob/33dba9518b4eb7fbc487fadc9718c408f95a826c/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L111-L234

To fix this, we should set a flag indicating whether the `CacheEntry.Value` was ever set. If it wasn't set, we shouldn't be committing the value into the cache.

cc @Tratcher

**Comment by Tratcher** [comment]
Design hindsight: Dispose was the wrong term, it should have been Commit or similar. Then Dispose could have been used for general cleanup without committing.

> To fix this, we should set a flag indicating whether the `CacheEntry.Value` was ever set. If it wasn't set, we shouldn't be committing the value into the cache.

👍

**Comment by eerhardt** [comment]
> Design hindsight: Dispose was the wrong term, it should have been named Commit or similar. Then Dispose could have been used for general cleanup without committing.

+💯 

> There's still some risk of an exception being thrown after the value was set, such as when setting up the change tokens, timeouts, etc..

Agreed, but that is more of an edge case. That won't happen when using the `GetOrCreate` methods, only if someone is calling `CreateEntry` themselves. To fix that, they can move `SetValue` to the last thing before calling `Dispose`.

One other risk is that if someone is relying on the behavior of not calling `SetValue` to mean "cache `null`". In that case, they would need to start calling `SetValue(null)`. But I think that is much less common than the mainline here.

**Comment by eerhardt** [comment]
Reopening to back port to 5.0.

**Comment by eerhardt** [comment]
Closing as the fix is now in the 5.0 branch.
